saw-core/src/SAWCore/Term/CtxTerm.hs

```diff
@@ -1120,100 +722,78 @@ instance UsesDataType (Bindings CtxTerm ctx as) where
+      i == length ps + length c1 + length c2 &&
+      isVarList ps c1 c2 ts
+    isVarList _ _ _ _ = False
 asCtorDTApp _ _ _ _ _ _ = Nothing
 
 
--- | Existential return type for 'asCtorArg'
-data ExCtorArg d ixs ctx =
-  forall a. ExCtorArg (CtorArg d ixs ctx (Typ a))
-
 -- | Check that an argument for a constructor has one of the allowed forms
-asCtorArg :: DataIdent d -> Bindings CtxTerm EmptyCtx params ->
-             Bindings CtxTerm (CtxInv params) ixs ->
-             InvBindings tp (CtxInv params) prevs ->
-             CtxTerm (CtxInv params <+> prevs) (Typ a) ->
-             Maybe (ExCtorArg d ixs (CtxInv params <+> prevs))
-asCtorArg d params dt_ixs prevs (ctxAsPiMulti ->
-                                 CtxMultiPi zs
-                                 (asCtorDTApp d params dt_ixs prevs zs ->
+asCtorArg :: Name -> [(LocalName, Term)] ->
+             [(LocalName, Term)] ->
+             [(LocalName, tp)] ->
+             Term ->
+             Maybe CtorArg
+asCtorArg d params dt_ixs prevs (asPiList ->
+                                 (zs,
+                                  asCtorDTApp d params dt_ixs prevs zs ->
                                   Just ixs))
   | not (usesDataType d zs)
-  = Just (ExCtorArg $ RecursiveArg zs ixs)
+  = Just (RecursiveArg zs ixs)
 asCtorArg d _ _ _ tp
   | not (usesDataType d tp)
-  = Just (ExCtorArg $ ConstArg tp)
+  = Just (ConstArg tp)
 asCtorArg _ _ _ _ _ = Nothing
 
--- | Existential return type of 'asPiCtorArg'
-data CtxPiCtorArg d ixs ctx =
-  forall a b .
-  CtxPiCtorArg LocalName (CtorArg d ixs ctx (Typ a))
-  (CtxTerm (ctx ::> a) (Typ b))
-
 -- | Check that a constructor type is a pi-abstraction that takes as input an
 -- argument of one of the allowed forms described by 'CtorArg'
-asPiCtorArg :: DataIdent d -> Bindings CtxTerm EmptyCtx params ->
-               Bindings CtxTerm (CtxInv params) ixs ->
-               InvBindings tp (CtxInv params) prevs ->
-               CtxTerm (CtxInv params <+> prevs) (Typ a) ->
-               Maybe (CtxPiCtorArg d ixs (CtxInv params <+> prevs))
-asPiCtorArg d params dt_ixs prevs (ctxAsPi ->
-                                   Just (CtxPi x
-                                         (asCtorArg d params dt_ixs prevs ->
-                                          Just (ExCtorArg arg)) rest)) =
-  Just $ CtxPiCtorArg x arg (castTopCtxElem rest)
-  where
-    castTopCtxElem :: CtxTerm (ctx ::> a1) b -> CtxTerm (ctx ::> a2) b
-    castTopCtxElem (CtxTerm t) = CtxTerm t
+asPiCtorArg :: Name -> [(LocalName, Term)] ->
+               [(LocalName, Term)] ->
+               [(LocalName, tp)] ->
+               Term ->
+               Maybe (LocalName, CtorArg, Term)
+asPiCtorArg d params dt_ixs prevs (asPi ->
+                                   Just (x,
+                                         asCtorArg d params dt_ixs prevs ->
+                                          Just arg, rest)) =
+  Just (x, arg, rest)
```

Comment:
There's quite a bit of rightward drift in this code. Consider factoring out some of this into a pattern guard:

```suggestion
asPiCtorArg d params dt_ixs prevs ty
  | Just (x, asCtorArg d params dt_ixs prevs -> Just arg, rest)) <- asPi ty =
    Just (x, arg, rest)
```

Reply:
Fixed in 8a513c1d5f55414109915264ecd4f09dd31d6d1a.